codegen/service/service_data.go

```diff
@@ -597,3 +683,252 @@ func requirements(m *design.MethodExpr) []*design.SecurityExpr {
+	rt := att.Type.(*design.ResultTypeExpr)
+	var (
+		views    []*ProjectData
+		validate string
+	)
+	vobj := design.AsObject(vatt.Type)
+	vtgt := "result"
+	validate = fmt.Sprintf("switch %s.View {", vtgt)
+	for _, view := range rt.Views {
+		obj := &design.Object{}
+		viewobj := view.Type.(*design.Object)
+		for _, n := range *viewobj {
+			if attr := vobj.Attribute(n.Name); attr != nil {
+				obj.Set(n.Name, attr)
+			}
+		}
+		valattr := &design.AttributeExpr{Type: obj}
+		if rt.Validation != nil {
+			valattr.Validation = rt.Validation.Dup()
+		}
+		validate += fmt.Sprintf("\ncase %q:\n%s", view.Name, codegen.RecursiveValidationCode(valattr, false, true, false, vtgt))
+		views = append(views, &ProjectData{
+			Name:        view.Name,
+			Description: view.Description,
+			Project:     projectToView(att, vatt, view, scope, viewspkg),
+		})
+	}
+	validate += "\n}"
+	return views, validate
+}
+
+// convertToResult converts the given viewed result type to result type.
+func convertToResult(vatt, res *design.AttributeExpr, viewspkg string, scope *codegen.NameScope) *InitData {
+	resrt := res.Type.(*design.ResultTypeExpr)
+	code, helpers := buildConstructorCode(vatt, res, "vres", "res", viewspkg, "", false, scope)
+	code += "\nreturn res"
+	rname := codegen.Goify(resrt.TypeName, true)
+	vname := "New" + rname
+	return &InitData{
+		VarName:     vname,
+		Description: fmt.Sprintf("%s converts viewed result type %s to result type %s.", vname, rname, rname),
+		Args: []*InitArgData{
+			{Name: "vres", Ref: scope.GoFullTypeRef(res, viewspkg)},
+		},
+		ReturnRef: scope.GoTypeRef(res),
+		Code:      code,
+		Helpers:   helpers,
+	}
+}
+
+// projectToView builds the constructor function to project the given
+// result type to the given viewed result type based on a view.
+func projectToView(res, vres *design.AttributeExpr, view *design.ViewExpr, scope *codegen.NameScope, viewspkg string) *InitData {
+	vresrt := vres.Type.(*design.ResultTypeExpr)
+	rname := scope.GoTypeName(res)
+	obj := &design.Object{}
+	tgt := &design.AttributeExpr{
+		Type: &design.UserTypeExpr{
+			AttributeExpr: &design.AttributeExpr{Type: obj},
+			TypeName:      vresrt.TypeName,
+		},
+	}
+	viewobj := view.Type.(*design.Object)
+	vresobj := design.AsObject(vres.Type)
+	// Select only the attributes from the given view
+	for _, n := range *viewobj {
+		attr := vresobj.Attribute(n.Name)
+		if attr == nil {
+			continue
+		}
+		// Add any specific view metadata from view attribute
+		if v, ok := n.Attribute.Metadata["view"]; ok {
+			if attr.Metadata == nil {
+				attr.Metadata = design.MetadataExpr{}
+			}
+			attr.Metadata["view"] = v
+		}
+		obj.Set(n.Name, attr)
+	}
+	code, helpers := buildConstructorCode(res, tgt, "res", "t", "", viewspkg, true, scope)
+	code += fmt.Sprintf("\nreturn &%s.%s{\n%s: t,\nView: %q,\n}", viewspkg, rname, vresrt.TypeName, view.Name)
```

Comment:
Would be nice to use templates instead of building the code inline here and below. Would make it a lot easier to grok.